draft-ietf-quic-version-negotiation.md

```diff
@@ -269,46 +333,103 @@ initially attempted by default, the client SHOULD cache that information and
 attempt the preferred version in subsequent connections.
 
 
-# Compatible Versions
+# Client Choice of Original Version
+
+The client's first flight SHOULD be sent using the version that the server is
+most likely to support (in the absence of other information, this will often be
+the oldest version the client supports).
+
+
+# Interaction with Retry
+
+QUIC version 1 features retry packets, which the server can send to validate
+the client's IP address before parsing the client's first flight. This impacts
+compatible version negotiation because a server who wishes to send a retry
+packet before parsing the client's first flight won't have parsed the client's
+handshake version information yet. If a future document wishes to define
+compatibility between two versions that support retry, that document MUST
+specify how version negotiation (both compatible and incompatible) interacts
+with retry during a handshake that requires both. For example, that could be
+accomplished by having the server send a retry packet first and validating the
+client's IP address before starting version negotiation and deciding whether to
+use compatible version negotiation on that connection (in that scenario the
+retry packet would be sent using the original version).
 
-Two versions of QUIC A and B are said to be "compatible" if a version A Initial
-can be used to negotiate version B and vice versa. The most common
-scenario is a sequence of versions 1, 2, 3, etc. in which all the
-Initial packets have the same basic structure but might include
-specific extensions (especially inside the crypto handshake)
-that are only meaningful in some subset of versions and are ignored
-in others. Note that it is not possible to add new frame types in
-Initial packets because QUIC frames do not use a self-describing
-encoding, so unrecognized frame types cannot be parsed or ignored (see the
-Extension Frames section of {{QUIC}}).
 
-When a new version of QUIC is defined, it is assumed to not be compatible
-with any other version unless otherwise specified. Implementations MUST NOT
-assume compatibility between version unless explicitly specified.
+# Interaction with 0-RTT
+
+QUIC version 1 allows sending data from the client to the server during the
+handshake, by using 0-RTT packets. If a future document wishes to define
+compatibility between two versions that support 0-RTT, that document MUST
+address the scenario where there are 0-RTT packets in the client's first
+flight. For example, this could be accomplished by defining which
+transformations are applied to 0-RTT packets. Alternatively, that document
+could specify that compatible version negotiation causes 0-RTT data to be
+rejected by the server.
```

Comment:
This is an interesting point that we didn't really consider in detail with TLS 1.3. However, I think it's off base for the same reason as above. We should just require them to be valid.